File: reports.pl
#!/usr/bin/perl

# reports.pl -- ArchivesSpace batch exports

use strict;
use warnings;
use LWP::UserAgent;
use JSON;
use Switch;
use YAML::XS 'LoadFile';
use open ':std', ':encoding(UTF-8)';
binmode(STDOUT, ":utf8");

require 'as_utils.pl';

# Initialize the user agent
my $ua = LWP::UserAgent->new;

# Load config file
my $config = LoadFile('config.yml');
my $url = $config->{url};
my $repo = $config->{repo};

my $json_path = $config->{json_path};
my $eac_path = $config->{eac_path};
my $ead_path = $config->{ead_path};
my $marc_path = $config->{marc_path};

# Initialize ArchivesSpace session (may make more sense to do this in a subroutine)
#my $session = &new_session($url);
#$ua->default_header('X-ArchivesSpace-Session' => $session);

# Choose the data model to work with
my $model = &select_data_model();

# Select the report to run based on the data model
my $report = &select_report($model);

my $ids;

# Choose the type of report to run, based on the data model
if($model eq "agents") {
  &execute_report($report, "$model/people");
  &execute_report($report, "$model/corporate_entities");
  &execute_report($report, "$model/families");
  &execute_report($report, "$model/software");
} else {
  &execute_report($report, $model);
}

sub select_report {
  my $report_type;
  switch($_[0]) {
    case /agents/ {
      print "Select report type:\n* (1) Generic JSON\n* (2) EAC-CPF\n> ";
      chomp($report_type = <STDIN>);
      switch($report_type) {
        case 1 { $report_type = "json"; }
        case 2 { $report_type = "eac"; }
        else {
          print "Invalid entry, try again.\n";
          $report_type = &select_report($_[0]);
        }
      }
    }
    case /resources/ {
      print "Select report type:\n* (1) Generic JSON\n* (2) EAD\n* (3) MARC\n> ";
      chomp($report_type = <STDIN>);
      switch($report_type) {
        case 1 { $report_type = "json"; }
        case 2 { $report_type = "ead"; }
        case 3 { $report_type = "marc"; }
        else {
          print "Invalid entry, try again.\n";
          $report_type = &select_report($_[0]);
        }
      }
    }
    # The other reports will go here, but I haven't written them yet 
    else {
      print "Select report type:\n* (1) Generic JSON\n> ";
      chomp($report_type = <STDIN>);
      switch($report_type) {
        case 1 { $report_type = "json"; }
        else {
          print "Invalid entry, try again.\n";
          $report_type = &select_report($_[0]);
        }
      }
    }
  }

  return $report_type;
}

sub execute_report {
  my $report = $_[0];
  my $model = $_[1];
  my($s, $l, $p);
  #my $session = "da40aefff8554d12a18ac3174e8d4cdb398b7688da2111d0d3ed20211b376e6c";

  if($_[2]) { $s = $_[2]; } else {
    ($s, $l, $p) = &new_session($url);
  }

  $ua->default_header('X-ArchivesSpace-Session' => $s);

  my $ids;
  switch($model) {
    case /subjects/ { 
      print "Grab $url/$model IDs... \n";
      $ids = $ua->get("$url/$model?all_ids=true");
    }
    case /^agents/ { 
      print "Grab $url/$model IDs... \n";
      $ids = $ua->get("$url/$model?all_ids=true"); 
    }
    else { 
      print "Grab $url/$repo/$model IDs... \n";
      $ids = $ua->get("$url/$repo/$model?all_ids=true");
    }
  }

  if($ids->code() eq 404) { die "Error: ".$ids->status_line().": $model\n"; } else {
    $ids = decode_json($ids->decoded_content);
    my $size = @$ids;
    my $i = 0;
    switch($report) {
      case /json/ {
        my $model_url;
        switch($model) {
          case /subjects/ { $model_url = "$url/$model"; }
          case /^agents/ { $model_url = "$url/$model"; }
          else { $model_url = "$url/$repo/$model"; }
        }
        $model =~ s/agents\///g;
        my $file_output = "$json_path/$model"."_report.json";
        if(-e $file_output) { unlink $file_output; }
        print "Writing report to $file_output... \n";
        open my $fh, '>>', $file_output or die "Error opening $file_output: $!\n";
        print $fh "{\"model\":\[";
        for my $id (@$ids) {
          $i++;
          my $record = &get_request("$model_url/$id", $s);
          print $fh $record;
          if($i < $size) { print $fh ","; }
        }
        print $fh "\]}";
        close $fh or die "Error closing $file_output: $!\n";
      }
      case /eac/ {
        $model =~ s/agents\///g;
        $model =~ s/software/softwares/;
        for my $id (@$ids) {
          my $model_url = "$url/$repo/archival_contexts/$model/$id.xml";
          my $file_output = "$eac_path/$model"."_"."$id"."_eac.xml";
          # this is because the DU EAC record takes too long to export
          if($model_url !~ m/corporate_entities\/1506/) {
            print "$model_url\n";
            my $record = &get_request($model_url, $s);
            if(-e $file_output) { unlink $file_output; }
            open my $fh, '>>', $file_output or die "Error opening $file_output: $!\n";
            print $fh $record;
            close $fh or die "Error closing $file_output: $!\n";
          }
        }
      }
      case /marc/ {
        for my $id (@$ids) {
          my $model_url = "$url/$repo/$model/marc21/$id.xml";
          #my $resource = &get_request("$url/$repo/$model/$id", $s);
          my $resource = decode_json(&get_request("$url/$repo/$model/$id", $s));
          my $num = lc($resource->{id_0});
          my $title = $resource->{title};
          print "Downloading MARC record for $num $title... \n";
          my $record = &get_request($model_url, $s);
          if($record) {
            my $file_output = "$marc_path/$num"."_marc.xml";
            print "Writing $file_output... \n";
            if(-e $file_output) { unlink $file_output; }
            open my $fh, '>>', $file_output or die "Error opening $file_output: $!\n";
            print $fh $record;
            close $fh or die "Error closing $file_output: $!\n";
          }
        }
      }
      case /ead/ {
        for my $id (@$ids) {
          my $model_url = "$url/$repo/resource_descriptions/$id.xml";
          my $resource = &get_request("$url/$repo/$model/$id", $s);
          $resource = decode_json($resource);
          my $num = lc($resource->{id_0});
          my $title = $resource->{title};
          print "Downloading EAD for $num $title... \n";
          my $ead = &get_request($model_url, $s);
          if($ead) {
            my $file_output = "$ead_path/$num"."_ead.xml";
            print "Writing $file_output... \n";
            if(-e $file_output) { unlink $file_output; }
            open my $fh, '>>', $file_output or die "Error opening $file_output: $!\n";
            print $fh $ead;
            close $fh or die "Error closing $file_output: $!\n";
          }
        }
      }
    }
  }
}

File: reports.rb
#!/usr/bin/env ruby

require 'io/console'
require 'net/http'
require 'uri'
require 'json'
require_relative 'astools'

@repo = "/repositories/2"
@path = "/home/kevin"

data_model_opts = {
  '1' => "resources",
  '2' => "archival_objects",
  '3' => "agents",
  '4' => "subjects",
  '5' => "digital_objects",
  '6' => "accessions",
  '7' => "top_containers",
  '8' => "container_profiles"
}

agent_types = ["people", "corporate_entities", "families", "software"]

def selector(type, opts)
  puts "Select a #{type}:"
  opts.each do |k,v|
    puts "* (#{k}) #{v}"
  end

  opt = gets.chomp
  if opts.has_key?(opt)
    return opts[opt]
  else
    puts "Invalid entry, try again."
    selector(type, opts)
  end
end

def get_report_type(model)
  opts = case model
  when "resources"
    { '1' => "json", '2' => "ead", '3' => "marc" }
  when "agents"
    { '1' => "json", '2' => "eac" }
  else
    { '1' => "json" }
  end

  selector("report type", opts)
end

def write_single_file(filename, data)
  File.delete?(filename) if File.exist?(filename)
  File.open(filename, 'w') {|f| f.write data}
end

def run_report(type, model)
  request_url = case model
  when /subjects/, /^agents/, /container_profiles/
    "/#{model}"
  else
    "#{@repo}/#{model}"
  end
  ids = ASTools::HTTP.get_json(request_url, 'all_ids' => true)
  model = model.gsub(/agents\//,'').gsub('software', 'softwares')

  case type

  # generic JSON output
  when "json"
    file_output = "#{@path}/#{model}_report.json"
    File.delete(file_output) if File.exist?(file_output)
    f = File.open(file_output, 'a')
    f.write("{\"#{model}\":[")
    ids.each_with_index do |id, i|
      print "Writing #{model} #{type} record #{i+1} of #{ids.length}... \r"
      json = ASTools::HTTP.get_json("#{request_url}/#{id}")
      f.write(json.to_json)
      f.write(",") unless i == ids.length - 1
    end
    f.write("]}")
    f.close unless f.nil?

  # Encoded Archival Context (EAC) output
  when "eac"
    url = "#{@repo}/archival_contexts/#{model}"
    ids.each_with_index do |id, i|
      print "Writing #{type} record #{i+1} of #{ids.length}... \r"
      data = ASTools::HTTP.get_data("#{request_url}/#{id}.xml")
      write_single_file("#{@path}/eac/#{model}_#{id}_eac.xml", data)
    end

  # MARCXML output
  when "marc"
    ids.each_with_index do |id, i|
      print "Writing #{type} record #{i+1} of #{ids.length}... \r"
      num = ASTools::HTTP.get_json("/#{@repo}/#{model}/#{id}")['id_0'].downcase
      data = ASTools::HTTP.get_data("/#{@repo}/#{model}/marc21/#{id}.xml")
      write_single_file("#{@path}/marc/#{num}_marc.xml", data)
    end

  # Encoded Archival Description output
  when "ead"
    ids.each_with_index do |id, i|
      print "Writing #{type} record #{i+1} of #{ids.length}... \r"
      num = ASTools::HTTP.get_json("#{@repo}/resources/#{id}")['id_0'].downcase
      data = ASTools::HTTP.get_data("#{@repo}/resource_descriptions/#{id}.xml")
      write_single_file("#{@path}/ead/#{num}_ead.xml", data)
    end
  end

  puts "\nDone."
end

ASTools::User.get_session

model = selector("data model", data_model_opts)
type = get_report_type(model)

if model == "agents"
  agent_types.each do |atype|
    run_report(type, "#{model}/#{atype}")
  end
else
  run_report(type, model)
end
